### PR TITLE
init export sql plugin

### PR DESCRIPTION
**Pull request by lastlink**
* support to generate sql from entity relation uml diagrams
* supports mysql, postgres, sqlserver, sqlite formats
* also supports foreign keys (might need to be tweaked was checking cell edges and style types to determine relationship).

## Examples:
* img: sql export example
* full source:
[uml_diag_entity_example.zip](https://github.com/jgraph/drawio/files/9748095/uml_diag_entity_example.zip)

related to #3083

### CODE REVIEW

**Comment by andrerav** [comment]
Hi @lastlink, I tried installing this plugin manually but I can't figure out how to actually use it. Could you clarify how the user should proceed after installing the plugin? Thanks! :)

**Comment by lastlink** [comment]
You can find the options under arrange > insert
  * from sql or to sql buttons.
  * Essentially certain tables will pick up table columns and types when exporting, and cardinality on the line arrows matter for foreign key generation.
  * composite create table statements will be generated on many to many, pk in left column defines primary keys in creation
drawio_sql_demo

**Comment by lastlink** [comment]
@andrerav

**Comment by andrerav** [comment]
Great, thank you @lastlink, will try it out :)

**Comment by lordmayzel** [comment]
@lastlink I am sorry, but where can I get the file to install the plugin manually?

**Comment by HaswinVidanage** [comment]
@lastlink Great plugin, I'm using your example file however when i insert a new set of tables with foreign keys the foreign key edges don't get created. Is there a workaround for this ? or am i missing something ?

**Comment by hillardjhune16** [comment]
Can someone help me in installing this plug-in? I cannot seem to make it work.

**Comment by noestl** [comment]
> Can someone help me in installing this plug-in? I cannot seem to make it work.

@hillardjhune16 To install it go to to "Extras>Plugins>SQL, then to use it "Arrange>Insert>From SQL"

**Comment by gillesouv** [comment]
I dont have the "To SQL " in my draw.io application. I use a mac and installes the plugin on DRAW.io

**Comment by dogfella** [comment]
This doesn't work.  It is exporting a bunch of css  and not including Key restraints and other vital info.  
I haven't bothered cutting and pasting the entire export, but here's a small example from your demo.  

`CREATE TABLE `Table5` (
	`t1_row2<span` style="font-weight: 400;">&nbsp;int</span>,
	`t5_row2<span` style="font-weight: 400;">&nbsp;int</span>
);
`

**Comment by dogfella** [comment]
@>[gillesouv](https://github.com/gillesouv) I dont have the "To SQL " in my draw.io application. I use a mac and installes the plugin on DRAW.io



> I dont have the "To SQL " in my draw.io application. I use a mac and installes the plugin on DRAW.io

@gillesouv I believe you might have to try "adding" that as a plugin.   From memory in the dropdown menu, I think it was simply the word "SQL", but don't hold my feet to the fire on that.

**Comment by rajpra8** [comment]
In desktop version, I see External Plugins: External plugins disabled. Is there way to enable?
  In web version, I tried to install https://raw.githubusercontent.com/lastlink/drawio/patch-sql-exporter/src/main/webapp/plugins/exportsql.js but still To Sql does not show up. Any extra step to install this plugin?
 When this can be merged to main repo?

**Comment by rajpra8** [comment]
@lastlink  I am able to enable the plugins by starting from commad line
/Applications/draw.io.app/Contents/MacOS/draw.io  --enable-plugins 
 but to Sql is not showing ddl. It comes up with empty pane when I have selected the node before calling to sql
  Am I missing some steps?

**Comment by dogfella** [comment]
> In desktop version, I see External Plugins: External plugins disabled. Is there way to enable? In web version, I tried to install https://raw.githubusercontent.com/lastlink/drawio/patch-sql-exporter/src/main/webapp/plugins/exportsql.js but still To Sql does not show up. Any extra step to install this plugin? When this can be merged to main repo?

@rajpra8 
I used a cloned version of this pull request by directly downloading the entire folder and running it locally on a local server.  It must be run on a server or it will fail due to Cors restrictions.  It was easy to get up and running this way.  I work on PCs, so I can't give much input about MACs.

However, I want to point out what I read below because I'm probably gonna bail on this.  After reading the following, it doesn't sound like draw.io will merge this.  

See:  
>>>>The level of complexity of this project means that even simple changes can break a lot of other moving parts. The amount of testing required is far more than it first seems. If we were to receive a PR, we'd have to basically throw it away and write it how we want it to be implemented. 
We are grateful for community involvement, bug reports, & feature requests. We do not wish to come off as anything but welcoming, however, we've made the decision to keep this project closed to contributions for the long term viability of the project.

<<<<
I have no idea what kind of database you are trying to run, but I'm sick of messing with third party half-baked hacks and software.   It is a total time suck and people online can't write complete sentences in the tech world to document their work or even provide legible and literate input.   They abbreviate sentences and absolutely will hardly write more than a sentence or two trying to communicate online.  It is a frickin' joke.  The era of text messaging has taken us back to the era of smoke signals from teepees!  

So, I just spent the last few hours getting some real database and ERD tools installed using:
1. Oracle SQL Developer Data Modeler 
and 
2. Oracle SQL Developer
 
They are both up and running here.   With the help of Third Party JDBC Drivers, I was able to get an ERD diagram running and connecting with a MySQL MariaDB.   

To achieve this, here is the JDBC driver I needed along with Oracle's two separate pieces of software that I identified above. https://dev.mysql.com/downloads/connector/j/ 
Here is a guide:    https://alvinbunk.wordpress.com/2017/06/29/using-oracle-sql-developer-to-connect-to-mysqlmariadb-databases/
If you are running from a localhost and not online, you will want to also check out this thread about using the third party JDBC drivers and how to get them to sing with your localhost database.  It worked like a champ.  
https://asktom.oracle.com/pls/apex/f?p=100:11:0::::P11_QUESTION_ID:9531897800346175195

>>>>For the connection I did:
jdbc url = jdbc:mysql://localhost:3306/mysql
driver class = com.mysql.jdbc.Driver
<<<<

And here is a screen capture of he diagram I just grabbed from a few tables in my database.  I haven't tested any Schema exporting or fiddled with the software yet, but I will dive in tomorrow.  At least it works, which is a lot more than you can say for a lot of other attempts I've made with other tools.  

erd

Note that 
1. Oracle SQL Developer Data Modeler does the heavy lifting creating the ERD 
and 
2. Oracle SQL Developer is more geared towards working directly with the database itself.  The former is the drawing tool, while the latter is the database connection and editing tool.  

Good luck.

**Comment by lastlink** [comment]
A big issue with the drawio extensions is the cors doesn't allow you to easily add external ones. The easiest way for this is to maybe combine the 2 sql import and sql export and make it so only 1 file needs to be changed for easier install. I don't really have time on my plate for this now. Also for the import the names have to match for the foreign keys lines to properly generate and vice versus.

**Comment by MzamoTembe** [comment]
How do you even install this

File: src/main/webapp/plugins/exportsql.js [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+            // foreign keys
+            let entityFKeys = this.relationships.filter((relation) => relation.entityB == entityKey);
+            if (entityFKeys.length > 0) {
+                for (let i = 0; i < entityFKeys.length; i++) {
+                    const fk = entityFKeys[i];
+                    const fkRelTxt = fk.roleA;
+                    // must match format "[..] to [..]"
+                    const keySplit = "] to [";
+                    if (fkRelTxt.indexOf(keySplit) != -1 &&
+                        fkRelTxt[0] == "[" &&
```
Comment (fahmisarker) [inline src/main/webapp/plugins/exportsql.js:121]: Use `===` instead of `==' ?

**Pull request closed without merging**